kimchi/src/folding/expressions.rs

```diff
@@ -0,0 +1,304 @@
+    fn default() -> Self {
+        Self {
+            degree_0: vec![],
+            degree_1: vec![],
+            degree_2: vec![],
+        }
+    }
+}
+pub fn extract_terms<C: Clone, F>(exp: FoldingExp<C>, terms: &mut Vec<Term<C>>, f: &F)
+where
+    F: Fn(Term<C>) -> Term<C>,
+{
+    use Degree::*;
+    match exp {
+        e @ FoldingExp::Constant(_) => {
+            let term = f(Term::new(e, Zero));
+            terms.push(term)
+        }
+        FoldingExp::Cell(col) => {
+            let degree = match col {
+                FoldingColumn::Witness(_) | FoldingColumn::WitnessExtended(_) => One,
+                _ => Zero,
+            };
+            let exp = FoldingExp::Cell(col);
+            let term = f(Term::new(exp, degree));
+            terms.push(term)
+        }
+        FoldingExp::Double(e) => {
+            let f = |t| {
+                let Term { exp, sign, degree } = t;
+                let exp = FoldingExp::Double(Box::new(exp));
+                f(Term { exp, sign, degree })
+            };
+            let f: Box<dyn Fn(Term<C>) -> Term<C>> = Box::new(f);
+            extract_terms(*e, terms, &f);
+        }
+        FoldingExp::Square(e) => {
+            let mut temp = Vec::with_capacity(4);
+            extract_terms(*e.clone(), &mut temp, f);
+            for term in temp {
+                let Term { exp, sign, degree } = term;
+                let exp = Box::new(exp);
+                let f = |t| {
+                    let Term {
+                        exp: exp2,
+                        sign: sign2,
+                        degree: degree2,
+                    } = t;
+                    let degree = &degree * &degree2;
+                    let sign = sign != sign2;
+                    //TODO: specialize square case
+                    let exp = FoldingExp::Mul(exp.clone(), Box::new(exp2));
+                    Term { exp, sign, degree }
+                };
+                let f: Box<dyn Fn(Term<C>) -> Term<C>> = Box::new(f);
+                extract_terms(*e.clone(), terms, &f);
+            }
+        }
+        FoldingExp::Add(e1, e2) => {
+            extract_terms(*e1, terms, f);
+            extract_terms(*e2, terms, f);
+        }
+        FoldingExp::Sub(e1, e2) => {
+            extract_terms(*e1, terms, f);
+            let f = |t: Term<C>| f(-t);
+            let f: Box<dyn Fn(Term<C>) -> Term<C>> = Box::new(f);
+            extract_terms(*e2, terms, &f);
+        }
+        FoldingExp::Mul(e1, e2) => {
+            let mut temp = Vec::with_capacity(4);
+            extract_terms(*e1, &mut temp, f);
+            for term in temp {
+                let Term { exp, sign, degree } = term;
+                let exp = Box::new(exp);
+                let f = |t| {
+                    let Term {
+                        exp: exp2,
+                        sign: sign2,
+                        degree: degree2,
+                    } = t;
+                    let degree = &degree * &degree2;
```

Comment:
why is the degree multiplied instead of added here?

Reply:
ditto